src/scanner.c

```diff
@@ -0,0 +1,403 @@
+
+#include <tree_sitter/parser.h>
+#include <ctype.h> // isspace()
+#include <string.h> // memcpy()
+
+enum TokenType {
+  NEWLINE,
+  SEMICOLON,
+  ELSE,
+  RAW_STRING_LITERAL,
+  OPEN_PAREN,
+  CLOSE_PAREN,
+  OPEN_BRACE,
+  CLOSE_BRACE,
+  OPEN_BRACKET,
+  CLOSE_BRACKET,
+  OPEN_BRACKET2,
+  CLOSE_BRACKET2
+};
+
+// ---------------------------------------------------------------------------------------
+// Temporary Stack structure until we can use the `<tree_sitter/array.h>` header from
+// tree sitter 1.0.0. Inspired from tree-sitter-julia.
+
+typedef char Scope;
+
+const Scope SCOPE_TOP_LEVEL = 0;
+const Scope SCOPE_BRACE = 1;
+const Scope SCOPE_PAREN = 2;
+const Scope SCOPE_BRACKET = 3;
+const Scope SCOPE_BRACKET2 = 4;
+
+// A `Stack` data structure for tracking the current `Scope`
+//
+// `SCOPE_TOP_LEVEL` is never actually pushed onto the stack. It is returned from
+// `stack_peek()` as a base case when `len = 0`. Note that in `stack_pop()` we still check
+// for `len > 0` before peeking to retain the invariant that we can't pop without
+// something on the stack.
+// 
+// This actually makes serialization/deserialization very simple. Even if we pushed an
+// initial `SCOPE_TOP_LEVEL` in the create hook, there is no guarantee that that will get
+// serialized (so the length of the stack won't be remembered) because the serialize hook
+// only runs when we accept a token from our external scanner. That would complicate the
+// deserialize hook by forcing us to differentiate between the cases of:
+// 1) A deserialization call restoring state from a previous serialization (len > 0).
+// 2) A deserialization call when there wasn't a previous serialization (len = 0), where
+//    we'd have to repush an initial `SCOPE_TOP_LEVEL`.
+typedef struct {
+  Scope *arr;
+  unsigned len;
+} Stack;
+
+static Stack *new_stack() {
+  Scope *arr = malloc(TREE_SITTER_SERIALIZATION_BUFFER_SIZE);
+  if (arr == NULL) exit(1);
+  Stack *stack = malloc(sizeof(Stack));
+  if (stack == NULL) exit(1);
+  stack->arr = arr;
+  stack->len = 0;
+  return stack;
+}
+
+static void stack_free(Stack *stack) {
+  free(stack->arr);
+  free(stack);
+}
+
+static void stack_push(Stack *stack, Scope scope) {
+  if (stack->len >= TREE_SITTER_SERIALIZATION_BUFFER_SIZE) exit(1);
```

Comment:
I'm hoping that 1024 nested scopes would be pathological 😭

Reply:
Might be worth doublechecking the behavior here if you try to open a document in Positron that contains 1024 `{` or something, just to make sure this doesn't get `ark` into a crash-loop (crashing and then restarting on every keystroke or something)?

Reply:
It did in fact cause an infloop due to the `exit(1)` when you do this, since that nukes the whole process.

The current `next` behavior is mildly better than that. It doesn't crash but it isn't respecting the `TREE_SITTER_SERIALIZATION_BUFFER_SIZE`, and is likely serializing way too much.

In https://github.com/r-lib/tree-sitter-r/pull/75/commits/10236eb2d2b19875aaa1b33a712d319b9aafb7b6 I implemented an alternative where:
- For `stack_push()`, if the stack is full then we return `false`, causing the caller to also return `false` from `scan()`
- For `stack_pop()`, if the stack is empty or the expected `scope` doesn't match the actually popped one, then we also propagate a `false` up to `scan()`

Ultimately this is our way of saying "the external scanner can't deal with this right now". That causes the internal scanner to kick in here instead, which will either recognize and handle the character or skip it if the character is not recognized, possibly resulting in an AST error down the line, but not one that causes a process crash.